include/cetl/variable_length_array.hpp

```diff
@@ -0,0 +1,1710 @@
+
+static_assert(
+    __cplusplus >= 201402L,
+    "Unsupported language: ISO C14, C++14, or a newer version of either is required to use the built-in VLA type");
+
+namespace cetl
+{
+
+///
+/// Minimal, generic container for storing Cyphal variable-length arrays. While this type shares similarities with
+/// std::vector it has several important properties that are different.
+/// First, the `max_size_max` maximum bound allows the array to enforce a maximum size whether or not the allocator
+/// properly implements the std::allocator_traits::max_size protocol (std::pmr::polymorphic_allocator, for example).
+/// This allows use of an allocator that is backed by statically allocated memory:
+/// @snippet{trimleft} example_05_variable_length_array_vs_vector.cpp example_exact_fit
+///
+/// Even if cetl::VariableLengthArray::reserve() is not used the implementation will still use less memory then its
+/// STL counterpart:
+/// @snippet{trimleft} example_05_variable_length_array_vs_vector.cpp example_tight_fit_0
+/// @snippet{trimleft} example_05_variable_length_array_vs_vector.cpp example_tight_fit_1
+///
+/// Finally, when exceptions are disabled the cetl::VariableLengthArray will not exhibit undefined behaviour like
+/// std::vector does:
+/// @snippet{trimleft} example_05_variable_length_array_vs_vector.cpp example_no_exceptions
+///
+/// @tparam  T           The type of elements in the array.
+/// @tparam Allocator    The allocator type to use for all allocations.
+///
+template <typename T, typename Allocator>
+class VariableLengthArray
+{
+public:
+    ///
+    /// STL-like declaration of pointer type.
+    ///
+    using pointer = typename std::add_pointer_t<T>;
+
+    ///
+    /// STL-like declaration of the iterator type.
+    ///
+    using iterator = typename std::add_pointer_t<T>;
+
+    ///
+    /// STL-like declaration of the const iterator type.
+    ///
+    using const_iterator = typename std::add_pointer_t<typename std::add_const_t<T>>;
+
+    ///
+    /// STL-like declaration of the container's storage type.
+    ///
+    using value_type = T;
+
+    ///
+    /// STL-like declaration of the container's difference type.
+    ///
+    using difference_type = std::ptrdiff_t;
+
+    ///
+    /// STL-like declaration of the container's size type.
+    ///
+    using size_type = std::size_t;
+
+    ///
+    /// Per the std::uses_allocator protocol.
+    ///
+    using allocator_type = typename std::allocator_traits<Allocator>::template rebind_alloc<T>;
+
+    /// Required constructor for std::uses_allocator protocol.
+    /// @param alloc Allocator to use for all allocations.
+    /// @param max_size_max Clamping value for the maximum size of this array. That is,
+    /// cetl::VariableLengthArray::max_size() will return
+    /// `std::min(max_size_max, std::allocator_traits<Allocator>::max_size(alloc))`
+    explicit constexpr VariableLengthArray(
+        const allocator_type& alloc,
+        size_type             max_size_max =
```

Comment:
These max_size_max calculations seem wrong.  Allocator max_size returns the maximum number of elements, not the maximum number of bytes (see https://en.cppreference.com/w/cpp/memory/allocator/max_size)

Reply:
Good catch. It's actually a bit more nuanced then that. The maxsize is actually the maximum diffval since you have to be able to iterate over all elements in a collection even if you can allocate more memory (theoretically, of course)

Reply:
Actually, now that I look at this it is correct. So, the max_size_max value allows a user to set an artificial ceiling. For example:

```
auto foo = VariableLengthArray(std::allocator(), 24);
```

since std::allocator doesn't implement max_size you can set a max_size_max  to implement a DSDL bounded-variable-length array. That this max size default is too large is okay because we always correctly calculate the value in the max_size() method.

Also, max_size_bytes == max_size_objects where sizeof(T) == 1.

Reply:
Hmm, there is some subtly here.  max_size() on allocator, allocator_traits, and vector, all represent the theoretical maximum number n of T elements (regardless as to whether there is that much memory available.)

If we followed that model then VariableLengthArray::max_size() should also represent the theoretical max and the impl should just be numeric_limits<size_type>::max() / sizeof(value_type).

Instead of max_size_max, I think the concept we should expose is capacity_max, passing that as a constructor arg and having a capacity_max() method.  This keeps VariableLengthArray::max_size() semantically the same as allocator/allocator_traits/vector and exposes VLA's unique new property, that of being bounded, as a new API (which the vector API has no way to represent.)

What do you think?